### ISSUE TITLE
App not able to recognize Salesforce Cordova plugin 

### ISSUE THREAD

**Issue by Vivekdeepak** [forcedotcom/SalesforceMobileSDK-Android#1190]
Please fill out the following details:

1. Version of Mobile SDK Used: 4.2
2. Issue found in Native App or Hybrid App: Hybrid App (Hybrid Local)
3. OS Version: 23
4. Device: 
5. Steps to reproduce: forcedroid create and the app does not show login page
6. Actual behavior: app starts without authentication
7. Expected Behavior: app should login to salesforce then show the data 
8. Error Log: com.salesforce.plugin.oauth class not found

**Comment by bhariharan** [comment]
@Vivekdeepak How did you create the app using ```forcedroid```? Could you post the entire console output of the creation process here?

**Comment by Vivekdeepak** [comment]
Yes,
I created the app using forcedroid, then build it using cordova build after that I opened into the emulator or the mobile it does not start with Salesforce login page, it shows the blank users page. I checked the logs and found different errors but only able to understand this com.salesforce.plugin.oauth. 

I have been building apps in past used Salesforce SDK 2.1,2.2 etc but now I have to upgrade all those apps for latest cordova version and android version.

**Comment by bhariharan** [comment]
Could you post your ```forcedroid create``` console output?

**Comment by Vivekdeepak** [comment]
07-13 15:34:39.050: W/ContextImpl(9151): Implicit intents with startService are not safe: Intent { act=tv.gamepop.home.utils.GuidanceScreenForAd (has extras) } com.android.server.am.ActivityStackSupervisor.bstShowAdGuidance:2895 com.android.server.am.ActivityStackSupervisor.bstSendTopActivityInfo:2779 com.android.server.am.ActivityStackSupervisor.startActivityLocked:1244 
07-13 15:34:39.050: I/ActivityManager(9151): START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10204000 cmp=com.ionic101.app/.MainActivity} from pid 9347
07-13 15:34:39.060: I/PGA(9151): PgaGetGlTransport: Initializing transport = 0
07-13 15:34:39.060: I/PGA(9151): Attempting to create new SOCKET connectionn pid = 9151, tid = 9366
07-13 15:34:39.060: E/GuidanceScreen(9369): Initiating download of new key mapping for package: com.ionic101.app
07-13 15:34:39.060: I/PGA(9151): New SOCKET connection: system_server (pid 9151, tid 9366)
07-13 15:34:39.090: W/audio_hw_primary(7897): out_write() limiting sleep time 112539 to 46439
07-13 15:34:39.090: I/dalvikvm-heap(9151): Grow heap (frag case) to 6.056MB for 124212-byte allocation
07-13 15:34:39.120: I/dalvikvm-heap(9151): Grow heap (frag case) to 6.130MB for 124212-byte allocation
07-13 15:34:39.160: W/audio_hw_primary(7897): out_write() limiting sleep time 65759 to 46439
07-13 15:34:39.160: E/Sensor-AccelerometerUI(9347): java.io.IOException: Connection refused
07-13 15:34:39.160: W/ContextImpl(9347): Implicit intents with startService are not safe: Intent { act=com.bluestacks.home.svc } android.content.ContextWrapper.startServiceAsUser:505 com.bluestacks.BstCommandProcessor.BstCommandLoop.sendStartServiceIntent:1190 com.bluestacks.BstCommandProcessor.BstCommandLoop.sendStartAppIntent:1265 
07-13 15:34:39.170: W/ContextImpl(9151): Implicit intents with startService are not safe: Intent { act=tv.gamepop.home.utils.GuidanceScreenForAd (has extras) } com.android.server.am.ActivityStackSupervisor.bstShowAdGuidance:2895 com.android.server.am.ActivityStackSupervisor.bstSendTopActivityInfo:2779 com.android.server.am.ActivityStack.resumeTopActivityLocked:1520 
07-13 15:34:39.190: I/ActivityManager(9151): Start proc com.ionic101.app for activity com.ionic101.app/.MainActivity: pid=10445 uid=10056 gids={50056, 3003}
07-13 15:34:39.200: I/PGA(9151): Attempting to create new SOCKET connectionn pid = 9151, tid = 9164
07-13 15:34:39.200: I/PGA(9151): New SOCKET connection: system_server (pid 9151, tid 9164)
07-13 15:34:39.260: E/Home.AppLaunchReceiver(9430): Initiating download of new key mapping for package: com.ionic101.app
07-13 15:34:39.270: I/Home.KeyMappingDownloader(9430): url: http://cdn.bluestacks.com/public/appsettings/cfg/com.ionic101.app.cfg
07-13 15:34:39.380: I/dalvikvm(10445): Could not find method android.app.Activity.checkSelfPermission, referenced from method org.apache.cordova.CordovaInterfaceImpl.hasPermission
07-13 15:34:39.380: W/dalvikvm(10445): VFY: unable to resolve virtual method 1: Landroid/app/Activity;.checkSelfPermission (Ljava/lang/String;)I
07-13 15:34:39.380: I/dalvikvm(10445): Could not find method android.app.Activity.requestPermissions, referenced from method org.apache.cordova.CordovaInterfaceImpl.requestPermission
07-13 15:34:39.380: W/dalvikvm(10445): VFY: unable to resolve virtual method 19: Landroid/app/Activity;.requestPermissions ([Ljava/lang/String;I)V
07-13 15:34:39.380: I/dalvikvm(10445): Could not find method android.app.Activity.requestPermissions, referenced from method org.apache.cordova.CordovaInterfaceImpl.requestPermissions
07-13 15:34:39.380: W/dalvikvm(10445): VFY: unable to resolve virtual method 19: Landroid/app/Activity;.requestPermissions ([Ljava/lang/String;I)V
07-13 15:34:39.390: I/chromium(10445): [INFO:library_loader_hooks.cc(112)] Chromium logging enabled: level = 0, default verbosity = 0
07-13 15:34:39.390: I/BrowserProcessMain(10445): Initializing chromium process, renderers=0
07-13 15:34:39.390: W/chromium(10445): [WARNING:proxy_service.cc(888)] PAC support disabled because there is no system implementation
07-13 15:34:39.390: I/PGA(10445): PgaGetGlTransport: Initializing transport = 0
07-13 15:34:39.390: I/PGA(10445): Attempting to create new SOCKET connectionn pid = 10445, tid = 10445
07-13 15:34:39.400: I/PGA(10445): New SOCKET connection: com.ionic101.app (pid 10445, tid 10445)
07-13 15:34:39.460: I/dalvikvm-heap(10445): Grow heap (frag case) to 2.991MB for 281892-byte allocation
07-13 15:34:39.480: I/dalvikvm(10445): Could not find method android.webkit.CookieManager.setAcceptThirdPartyCookies, referenced from method org.apache.cordova.engine.SystemCookieManager.<init>
07-13 15:34:39.480: W/dalvikvm(10445): VFY: unable to resolve virtual method 149: Landroid/webkit/CookieManager;.setAcceptThirdPartyCookies (Landroid/webkit/WebView;Z)V
07-13 15:34:39.480: I/dalvikvm(10445): Could not find method android.webkit.CookieManager.flush, referenced from method org.apache.cordova.engine.SystemCookieManager.flush
07-13 15:34:39.480: W/dalvikvm(10445): VFY: unable to resolve virtual method 144: Landroid/webkit/CookieManager;.flush ()V
07-13 15:34:39.480: W/dalvikvm(10445): VFY: unable to find class referenced in signature (Landroid/webkit/ClientCertRequest;)
07-13 15:34:39.480: W/dalvikvm(10445): VFY: unable to find class referenced in signature (Landroid/webkit/ClientCertRequest;)
07-13 15:34:39.480: I/dalvikvm(10445): Could not find method android.webkit.WebViewClient.onReceivedClientCertRequest, referenced from method org.apache.cordova.engine.SystemWebViewClient.onReceivedClientCertRequest
07-13 15:34:39.480: W/dalvikvm(10445): VFY: unable to resolve virtual method 202: Landroid/webkit/WebViewClient;.onReceivedClientCertRequest (Landroid/webkit/WebView;Landroid/webkit/ClientCertRequest;)V
07-13 15:34:39.480: W/dalvikvm(10445): VFY: unable to find class referenced in signature (Landroid/webkit/PermissionRequest;)
07-13 15:34:39.480: I/dalvikvm(10445): Could not find method android.webkit.PermissionRequest.getResources, referenced from method org.apache.cordova.engine.SystemWebChromeClient.onPermissionRequest
07-13 15:34:39.480: W/dalvikvm(10445): VFY: unable to resolve virtual method 160: Landroid/webkit/PermissionRequest;.getResources ()[Ljava/lang/String;
07-13 15:34:39.480: I/dalvikvm(10445): Could not find method android.webkit.WebChromeClient$FileChooserParams.createIntent, referenced from method org.apache.cordova.engine.SystemWebChromeClient.onShowFileChooser
07-13 15:34:39.480: W/dalvikvm(10445): VFY: unable to resolve virtual method 165: Landroid/webkit/WebChromeClient$FileChooserParams;.createIntent ()Landroid/content/Intent;
07-13 15:34:39.530: W/AwContents(10445): nativeOnDraw failed; clearing to background color.
07-13 15:34:39.580: W/AwContents(10445): nativeOnDraw failed; clearing to background color.
07-13 15:34:39.580: I/ActivityManager(9151): Displayed com.ionic101.app/.MainActivity: +410ms
07-13 15:34:39.880: I/chromium(10445): [INFO:async_pixel_transfer_manager_android.cc(56)] Async pixel transfers not supported
07-13 15:34:39.960: I/chromium(10445): [INFO:async_pixel_transfer_manager_android.cc(56)] Async pixel transfers not supported
07-13 15:34:40.050: I/chromium(10445): [INFO:CONSOLE(59)] "Uncaught module com.salesforce.plugin.oauth not found", source: file:///android_asset/www/cordova.js (59)``

**Comment by bhariharan** [comment]
The output from app creation, not the ```logcat``` output.

**Comment by Vivekdeepak** [comment]
C:\MobileApps\MobileTraining>forcedroid create
Enter your application type (native, react_native, hybrid_remote, or hybrid_local): hybrid_local
hybrid_local
Enter your application name: IonicApp102
Enter the target directory of your app (must be an existing empty folder): C:\MobileApps\MobileTraining\IonicApp101
Enter the package name for your app (com.mycompany.my_app): com.demo1.app
Creating a new cordova project.
C:\MobileApps\MobileTraining\IonicApp101\IonicApp102 C:\MobileApps\MobileTraining
Adding android project...
Creating Cordova project for the Android platform:
        Path: platforms\android
        Package: com.demo1.app
        Name: IonicApp102
        Activity: MainActivity
        Android target: android-23
Android project created with cordova-android@5.0.0
Discovered plugin "cordova-plugin-whitelist" in config.xml. Installing to the project
Fetching plugin "cordova-plugin-whitelist@1" via npm
Installing "cordova-plugin-whitelist" for android

               This plugin is only applicable for versions of cordova-android greater than 4.0. If you have a previous platform version, you do *not* need this plugin since the whitelist will be built in.

Fetching plugin "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" via git clone
Repository "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" checked out to git ref "v4.1.1".
Installing "com.salesforce" for android
Dependent plugin "cordova-plugin-whitelist" already installed on android.
Fetching plugin "cordova-plugin-device" via npm
Installing "cordova-plugin-device" for android
Fetching plugin "phonegap-plugin-push" via npm
Installing "phonegap-plugin-push" for android
Failed to install 'phonegap-plugin-push':Error: Variable(s) missing: SENDER_ID
    at C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\cordova-lib\src\plugman\install.js:343:23
    at _fulfilled (C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q.js:787:54)
    at self.promiseDispatch.done (C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q.js:816:30)
    at Promise.promise.promiseDispatch (C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q.js:749:13)
    at C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q.js:557:44
    at flush (C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q.js:108:17)
    at doNTCallback0 (node.js:419:9)
    at process._tickCallback (node.js:348:13)
Error: Variable(s) missing: SENDER_ID
Failed to install 'com.salesforce':Error: Variable(s) missing: SENDER_ID
    at C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\cordova-lib\src\plugman\install.js:343:23
    at _fulfilled (C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q.js:787:54)
    at self.promiseDispatch.done (C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q.js:816:30)
    at Promise.promise.promiseDispatch (C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q.js:749:13)
    at C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q.js:557:44
    at flush (C:\Users\vivek\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q.js:108:17)
    at doNTCallback0 (node.js:419:9)
    at process._tickCallback (node.js:348:13)
C:\MobileApps\MobileTraining

Your application project is ready in C:\MobileApps\MobileTraining\IonicApp101.

To use your new application in Android Studio, do the following:
   - Launch Android Studio and select `Import project (Eclipse ADT, Gradle, etc.)` from the Welcome screen
   - Navigate to the C:\MobileApps\MobileTraining\IonicApp101/IonicApp102/platforms/android folder, select it and click `Ok`
   - From the dropdown that displays the available targets, choose the sample app you want to run and click the play button

Before you ship, make sure to plug your OAuth Client ID,
Callback URI, and OAuth Scopes into www/bootconfig.json

C:\MobileApps\MobileTraining>cd IonicApp101

C:\MobileApps\MobileTraining\IonicApp101>cd IonicApp102

C:\MobileApps\MobileTraining\IonicApp101\IonicApp102>cordova plugin list
com.salesforce 4.1.1 "SalesforceMobileSDK Plugins"
cordova-plugin-device 1.1.2 "Device"
cordova-plugin-whitelist 1.2.2 "Whitelist"
phonegap-plugin-push 1.7.4 "PushPlugin"

C:\MobileApps\MobileTraining\IonicApp101\IonicApp102>

**Comment by Vivekdeepak** [comment]
I also tried the other way
creating first cordova app then adding Salesforce Cordova plugin still does not work

**Comment by bhariharan** [comment]
@Vivekdeepak You're using ```forcedroid 4.1.1```, not ```forcedroid 4.2.0```, which is the latest. From your console logs:

```
Fetching plugin "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" via git clone
Repository "https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin" checked out to git ref "v4.1.1".
```

The problem here is the incorrect version of ```PushPlugin``` being used. This is fixed in ```forcedroid 4.2.0```. From your console logs:

```
Error: Variable(s) missing: SENDER_ID
Failed to install 'com.salesforce':Error: Variable(s) missing: SENDER_ID
```

Please grab the latest from [here](https://www.npmjs.com/package/forcedroid) and run ```forcedroid create```. That should work for you.
